cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -117,6 +120,130 @@ inline std::shared_ptr<std::unordered_map<std::pair<StreamId, VersionId>, AtomKe
+            load_param_.load_from_time_ = std::min(load_param_.load_from_time_.value(), timestamp_query.timestamp_);
+            break;
+        case LoadType::LOAD_DOWNTO:
+        case LoadType::LOAD_UNDELETED:
+            load_param_ = LoadParameter{LoadType::LOAD_UNDELETED};
+            break;
+        default:
+            util::raise_rte("Unexpected load state {} applying specific version query", load_param_.load_type_);
+        }
+    }
+
+    void do_react(const pipelines::SnapshotVersionQuery&) {
+        util::raise_rte("Snapshot not currently supported in generic batch read version");
+    }
+};
+
+inline std::optional<AtomKey> get_key_for_version_query(const std::shared_ptr<VersionMapEntry>& version_map_entry, const pipelines::VersionQuery& version_query) {
+    return util::variant_match(version_query.content_,
+        [&version_map_entry] (const pipelines::SpecificVersionQuery& specific_version) {
+        return find_index_key_for_version_id(specific_version.version_id_, version_map_entry);
+    },
+    [&version_map_entry] (const pipelines::TimestampVersionQuery& timestamp_version) {
+       return find_index_key_for_version_timestamp(timestamp_version.timestamp_, version_map_entry);
+    },
+    [&version_map_entry] (const std::monostate&) {
+       return version_map_entry->get_first_index(false);
+       },
+    [] (const auto&) -> std::optional<AtomKey> {
+        util::raise_rte("Unsupported version query type");
+    });
+}
+
+inline std::vector<folly::Future<std::optional<AtomKey>>> batch_get_versions(
+    const std::shared_ptr<Store>& store,
+    const std::shared_ptr<VersionMap>& version_map,
+    const std::vector<StreamId>& symbols,
+    const std::vector<pipelines::VersionQuery>& version_queries) {
+    ARCTICDB_SAMPLE(BatchGetVersion, 0)
+    util::check(symbols.size() == version_queries.size(), "Symbol and version query list mismatch: {} != {}", symbols.size(), version_queries.size());
+
+    robin_hood::unordered_flat_map<StreamId, StreamVersionData> version_data;
+    for(const auto& symbol : folly::enumerate(symbols)) {
+        auto it = version_data.find(*symbol);
+        if(it == version_data.end())
+            version_data.insert(robin_hood::pair<StreamId, StreamVersionData>(*symbol, StreamVersionData{version_queries[symbol.index]}));
+        else
+            it->second.react(version_queries[symbol.index]);
+    }
+
+    using SplitterType = folly::FutureSplitter<std::shared_ptr<VersionMapEntry>>;
+    robin_hood::unordered_flat_map<StreamId, SplitterType> shared_futures;
+    std::vector<folly::Future<std::optional<AtomKey>>> output;
+    output.reserve(symbols.size());
+    for(const auto& symbol : folly::enumerate(symbols)) {
+        const auto it = version_data.find(*symbol);
+        util::check(it != version_data.end(), "Missing version data for symbol {}", *symbol);
+        auto version_entry_fut = folly::Future<std::shared_ptr<VersionMapEntry>>::makeEmpty();
```

Comment:
done

Reply:
Actually, it your proposal was not compiling. Future needs to be initialised, so. considering it is more like a nicknack, I will leave it as it was, since I am already explicating the type in the initialisation.